src/axom/quest/tests/quest_sampling_shaper.cpp

```diff
@@ -0,0 +1,868 @@
+    this->runShaping(shape_file.getFileName(), initialGridFunctions);
+
+    // check that the result has a volume fraction field associated with the circle material
+    constexpr double expected_hole_area = .5 * .5 * M_PI;
+    constexpr double expected_disk_area = M_PI - expected_hole_area;
+    const auto range = this->meshBoundingBox().range();
+    const double expected_bg_area = range[0] * range[1] - M_PI;
+
+    this->checkExpectedVolumeFractions("disk", expected_disk_area);
+    this->checkExpectedVolumeFractions("hole", expected_hole_area);
+    this->checkExpectedVolumeFractions("void", expected_bg_area);
+  }
+
+  // clean up data collection
+  for(const auto& name :
+      {"vol_frac_void", "vol_frac_hole", "vol_frac_disk", "init_vf_bg"})
+  {
+    this->getDC().DeregisterField(name);
+  }
+
+  // Set background and inner hole materials to 'void'
+  {
+    ScopedTemporaryFile shape_file(axom::fmt::format("{}.yaml", testname),
+                                   axom::fmt::format(shape_template,
+                                                     contour_file.getFileName(),
+                                                     "void",
+                                                     "disk",
+                                                     "void"));
+
+    // Create an initial background material set to 1 everywhere
+    std::map<std::string, mfem::GridFunction*> initialGridFunctions;
+    {
+      auto* vf = this->registerVolFracGridFunction("init_vf_bg");
+      this->initializeVolFracGridFunction(
+        vf,
+        [](int, const Point2D&, int) -> double { return 1.; });
+      initialGridFunctions["void"] = vf;
+    }
+
+    this->validateShapeFile(shape_file.getFileName());
+    this->runShaping(shape_file.getFileName(), initialGridFunctions);
+
+    // check that the result has a volume fraction field associated with the circle material
+    constexpr double expected_disk_area = M_PI - .5 * .5 * M_PI;
+    const auto range = this->meshBoundingBox().range();
+    const double expected_void_area = range[0] * range[1] - expected_disk_area;
+
+    this->checkExpectedVolumeFractions("disk", expected_disk_area);
+    this->checkExpectedVolumeFractions("void", expected_void_area);
+  }
+
+  if(very_verbose_output)
+  {
+    this->getDC().Save(testname, axom::sidre::Group::getDefaultIOProtocol());
+  }
+}
+
+TEST_F(SamplingShaperTest, preshaped_materials)
+{
+  using Point2D = typename SamplingShaperTest::Point2D;
+
+  const std::string& testname =
+    ::testing::UnitTest::GetInstance()->current_test_info()->name();
+
+  const std::string shape_template = R"(
+dimensions: 2
+units: cm
+
+shapes:
+- name: background
+  material: {0}
+  geometry:
+    format: none
+# Left replaces background void
+- name: left_side
+  material: {1}
+  geometry:
+    format: none
+# Odd cells replace background void, but not left
+- name: odd_cells
+  material: {2}
+  geometry:
+    format: none
+  does_not_replace: [{1}]
+)";
```

Comment:
Note: We're using `geometry/format: none` to indicate that a "shape" in the list does not have geometric data, but that we want it to participate in the replacement rules.

Another possibility might be `geometry/format: preshaped` to indicate that an initial volume fraction has been supplied for the corresponding material. This terminology might be a bit more awkward (?), since even after a "shape" has been preshaped for a material, we can still supply further shapes for that material.

Thoughts and/or suggestions ?